### ISSUE TITLE
`enum.IntFlag` type narrowed too much when left operand is of type `int` and right operand is a canonical flag

### ISSUE THREAD

**Issue by jonathandung** [astral-sh/ty#3765]
### Summary

[Playground](https://play.ty.dev/8f250914-79e5-473c-aba1-5fd3eb725359)

### Version

ty 0.0.49 (b669da87d 2026-06-11)

**Comment by jonathandung** [comment]
In the above example, I expected the revealed type to be `Flags` rather than `Literal[Flags.A]`, because it does not hold that `int() & Flags.A == Flags.A`.

**Comment by jonathandung** [comment]
The stubs in typeshed are as follows:

```python
class IntFlag(int, ReprEnum, Flag, boundary=KEEP):  # type: ignore[misc]  # complaints about incompatible bases
    def __new__(cls, value: int) -> Self: ...
    def __or__(self, other: int) -> Self: ...
    def __and__(self, other: int) -> Self: ...
    def __xor__(self, other: int) -> Self: ...
    def __invert__(self) -> Self: ...
    __ror__ = __or__
    __rand__ = __and__
    __rxor__ = __xor__
```

It may be the `Self` return value in `__rand__` triggering this fault, because the type of `Flags.A`, even in this context, is `Literal[Flags.A]` and not `Flags`. I will see whether this issue should be moved to typeshed after testing with other type checkers.

**Comment by jonathandung** [comment]
Correct:

- [mypy](https://gist.github.com/mypy-play/0c29d5580d780221622b7d8af6221996)

Wrong (revealed as `int` rather than `Flags`:

- [pyright](https://pyright-play.net/?code=GYJw9gtgBApgdgV2gSwgBzCALlAhgrMAGigEk4sAxAG1wHMBYAKAGNaBndqG%2B9gCnJVadAJQAuZlClQAglAC8eAmD4jJ0gEIKlhVcxAwAbjFzUA%2BlgCeaGH2QVVUAGTdh7AHQyRQA)
- [basedpyright](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBApgdgV2gSwgBzCALlAhgrMAGigEk4sAxAG1wHMBYAKAGNaBndqG%2B9gCnJVadAJQAuZlClQAglAC8eAmD4jJ0gEIKlhVcxAwAbjFzUA%2BlgCeaGH2QVVUAGTdh7AHQyRQA)
- [pyrefly](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeS4ATrgLYAEM6ArvRDcbpQC52pNe4ANHQCS6LgDEoqAOYAddAGNpcOHSmy4ACjGTpMgJSIFdU3QCCdALy9%2BuLQZNmAQtdsCHCyjABuMVFAA%2BlykxDBaEOIOdABk6vpwhOYGIIIgZN5gUKSEAjRQFADEdAAKpJnZdGhYePh0irjokDJMlKhcEI2ECsUAyjAwdAAWXFzEcIgA9JMZMFk5nDKTjJOYuIpwkw1NEC1tHY2TdGCcvD6o0KjYsPWNza3tneh0uMQH6IkKZFxDjQC0fkocCebjkIAAzIQAIwAJjBCgA2jBKNQgQBdBRMdBsDjcGCYP6YCDeRQdPxuADku3QnBgFIUNK4f28AEcmMT8X8ANYwUh-VCKRTwNQ2CkAd1QlHQ9PQIAAvmkBWSYBJoDAKNUcAQSOQ5UA)
- zuban

**Comment by sharkdp** [comment]
We do not support `enum.Flag` yet, please see https://github.com/astral-sh/ty/issues/876
